File: src/main/java/at/bitfire/ical4android/iCalendar.java
/*
 * Copyright (c) 2013 – 2015 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 */

package at.bitfire.ical4android;


import net.fortuna.ical4j.data.CalendarBuilder;
import net.fortuna.ical4j.data.CalendarParserFactory;
import net.fortuna.ical4j.data.ParserException;
import net.fortuna.ical4j.model.DateTime;
import net.fortuna.ical4j.model.Dur;
import net.fortuna.ical4j.model.Parameter;
import net.fortuna.ical4j.model.ParameterFactory;
import net.fortuna.ical4j.model.ParameterFactoryImpl;
import net.fortuna.ical4j.model.ParameterFactoryRegistry;
import net.fortuna.ical4j.model.PropertyFactoryRegistry;
import net.fortuna.ical4j.model.component.VAlarm;
import net.fortuna.ical4j.model.component.VTimeZone;
import net.fortuna.ical4j.model.property.DateProperty;
import net.fortuna.ical4j.model.property.ProdId;
import net.fortuna.ical4j.util.CompatibilityHints;
import net.fortuna.ical4j.util.Strings;
import net.fortuna.ical4j.util.TimeZones;

import java.io.IOException;
import java.io.StringReader;
import java.net.URISyntaxException;
import java.util.TimeZone;
import java.util.UUID;
import java.util.logging.Level;

import lombok.Getter;
import lombok.NonNull;


public class iCalendar {
    // static ical4j initialization
    static {
        CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_RELAXED_UNFOLDING, true);
        CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_RELAXED_PARSING, true);
        CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_OUTLOOK_COMPATIBILITY, true);
    }

    public static ProdId prodId = new ProdId("+//IDN bitfire.at//ical4android");

    public String uid;
    public Integer sequence;


    protected void generateUID() {
        uid = UUID.randomUUID().toString();
    }


    // time zone helpers

    protected static boolean isDateTime(DateProperty date) {
        return date != null && date.getDate() instanceof DateTime;
    }

    /**
     * Returns the time-zone ID for a given date-time, or TIMEZONE_UTC for dates (without time).
     * TIMEZONE_UTC is also returned for DATE-TIMEs in UTC representation.
     *
     * @param date DateProperty (DATE or DATE-TIME) whose time-zone information is used
     */
    protected static String getTzId(DateProperty date) {
        if (isDateTime(date) && !date.isUtc() && date.getTimeZone() != null)
            return date.getTimeZone().getID();
        else
            return TimeZones.UTC_ID;
    }

    /**
     * Ensures that a given DateProperty has a time zone with an ID that is available in Android.
     *
     * @param date DateProperty to validate. Values which are not DATE-TIME will be ignored.
     */
    protected static void validateTimeZone(DateProperty date) {
        if (isDateTime(date)) {
            final TimeZone tz = date.getTimeZone();
            if (tz == null)
                return;
            final String tzID = tz.getID();
            if (tzID == null)
                return;

            String deviceTzID = DateUtils.findAndroidTimezoneID(tzID);
            if (!tzID.equals(deviceTzID))
                date.setTimeZone(DateUtils.tzRegistry.getTimeZone(deviceTzID));
        }
    }

    /**
     * Takes a string with a timezone definition and returns the time zone ID.
     *
     * @param timezoneDef time zone definition (VCALENDAR with VTIMEZONE component)
     * @return time zone id (TZID)  if VTIMEZONE contains a TZID,
     * null                 otherwise
     */
    public static String TimezoneDefToTzId(@NonNull String timezoneDef) {
        try {
            CalendarBuilder builder = new CalendarBuilder();
            net.fortuna.ical4j.model.Calendar cal = builder.build(new StringReader(timezoneDef));
            VTimeZone timezone = (VTimeZone) cal.getComponent(VTimeZone.VTIMEZONE);
            if (timezone != null && timezone.getTimeZoneId() != null)
                return timezone.getTimeZoneId().getValue();
        } catch (IOException|ParserException e) {
            Constants.log.log(Level.SEVERE, "Can't understand time zone definition", e);
        }
        return null;
    }


    // misc. iCalendar helpers

    protected static int alarmMinBefore(VAlarm alarm) {
        int minutes = 0;
        if (alarm.getTrigger() != null) {
            Dur duration = alarm.getTrigger().getDuration();
            if (duration != null) {
                // negative value in TRIGGER means positive value in Reminders.MINUTES and vice versa
                minutes = -(((duration.getWeeks() * 7 + duration.getDays()) * 24 + duration.getHours()) * 60 + duration.getMinutes());
                if (duration.isNegative())
                    minutes *= -1;
            }
        }
        return minutes;
    }


    // ical4j helpers and extensions

    private static final ParameterFactoryRegistry parameterFactoryRegistry = new ParameterFactoryRegistry();

    static {
        parameterFactoryRegistry.register(Email.PARAMETER_NAME, Email.FACTORY);
    }

    protected static CalendarBuilder calendarBuilder() {
        return new CalendarBuilder(
                CalendarParserFactory.getInstance().createParser(),
                new PropertyFactoryRegistry(), parameterFactoryRegistry, DateUtils.tzRegistry);
    }


    public static class Email extends Parameter {
        /* EMAIL property for ATTENDEE properties, as used by iCloud:
           ATTENDEE;EMAIL=bla@domain.tld;/path/to/principal
        */
        public static final ParameterFactory FACTORY = new Factory();

        public static final String PARAMETER_NAME = "EMAIL";
        @Getter private String value;

        protected Email() {
            super(PARAMETER_NAME, ParameterFactoryImpl.getInstance());
        }

        public Email(String aValue) {
            super(PARAMETER_NAME, ParameterFactoryImpl.getInstance());
            value = Strings.unquote(aValue);
        }

        public static class Factory implements ParameterFactory {
            @Override
            public Parameter createParameter(String value) throws URISyntaxException {
                return new Email(value);
            }

            @Override
            public boolean supports(String name) {
                return false;
            }
        }
    }
}

File: src/main/java/at/bitfire/ical4android/iCalendar.kt
/*
 * Copyright © Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.ical4android

import net.fortuna.ical4j.data.CalendarBuilder
import net.fortuna.ical4j.data.CalendarParserFactory
import net.fortuna.ical4j.data.ParserException
import net.fortuna.ical4j.model.*
import net.fortuna.ical4j.model.component.VAlarm
import net.fortuna.ical4j.model.component.VTimeZone
import net.fortuna.ical4j.model.property.DateProperty
import net.fortuna.ical4j.model.property.ProdId
import net.fortuna.ical4j.util.CompatibilityHints
import net.fortuna.ical4j.util.Strings
import java.io.StringReader
import java.net.URISyntaxException
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

open class iCalendar {

    var uid: String? = null
    var sequence: Int? = null

    companion object {
        // static ical4j initialization
        init {
            Constants.log.info("Enabling ical4j releaxed unfolding, relaxed parsing, Outlook compatibility")
            CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_RELAXED_UNFOLDING, true)
            CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_RELAXED_PARSING, true)
            CompatibilityHints.setHintEnabled(CompatibilityHints.KEY_OUTLOOK_COMPATIBILITY, true)

            // reduce verbosity of those two loggers
            org.slf4j.LoggerFactory.getLogger(net.fortuna.ical4j.data.CalendarParserImpl::class.java)
            Logger.getLogger(net.fortuna.ical4j.data.CalendarParserImpl::class.java.name).level = Level.CONFIG
            org.slf4j.LoggerFactory.getLogger(net.fortuna.ical4j.model.Recur::class.java)
            Logger.getLogger(net.fortuna.ical4j.model.Recur::class.java.name).level = Level.CONFIG
        }

        var prodId = ProdId("+//IDN bitfire.at//ical4android")

        private val parameterFactoryRegistry = ParameterFactoryRegistry()
        init {
            parameterFactoryRegistry.register(Email.PARAMETER_NAME, Email.FACTORY)
        }

        private val propertyFactoryRegistry = PropertyFactoryRegistry()
        init {
            propertyFactoryRegistry.register(Color.PROPERTY_NAME, Color.FACTORY)
        }

        @JvmStatic
        protected fun calendarBuilder() = CalendarBuilder(
                CalendarParserFactory.getInstance().createParser(),
                propertyFactoryRegistry, parameterFactoryRegistry, DateUtils.tzRegistry)


        // time zone helpers

        @JvmStatic
        fun isDateTime(date: DateProperty?) = date != null && date.date is DateTime

        /**
         * Takes a string with a timezone definition and returns the time zone ID.
         * @param timezoneDef time zone definition (VCALENDAR with VTIMEZONE component)
         * @return time zone id (TZID) if VTIMEZONE contains a TZID, null otherwise
         */
        @JvmStatic
        fun TimezoneDefToTzId(timezoneDef: String): String? {
            try {
                val builder = CalendarBuilder()
                val cal = builder.build(StringReader(timezoneDef))
                val timezone = cal.getComponent(VTimeZone.VTIMEZONE) as VTimeZone?
                timezone?.timeZoneId?.let { return it.value }
            } catch (e: ParserException) {
                Constants.log.log(Level.SEVERE, "Can't understand time zone definition", e)
            }
            return null
        }


        // misc. iCalendar helpers

        @JvmStatic
        internal fun alarmMinBefore(alarm: VAlarm): Int {
            var minutes = 0
            alarm.trigger?.duration?.let { duration ->
                // negative value in TRIGGER means positive value in Reminders.MINUTES and vice versa
                minutes = -(((duration.weeks * 7 + duration.days) * 24 + duration.hours) * 60 + duration.minutes + duration.seconds/60)
                if (duration.isNegative)
                    minutes *= -1
            }
            return minutes
        }

    }


    protected fun generateUID() {
        uid = UUID.randomUUID().toString()
    }


    override fun toString() = MiscUtils.reflectionToString(this)


    // ical4j helpers and extensions

    /** COLOR property for VEVENT components [RFC 7986 5.9 COLOR] */
    class Color(
            var value: EventColor? = null
    ): Property(PROPERTY_NAME, PropertyFactoryImpl.getInstance()) {
        companion object {
            val FACTORY = Factory()
            val PROPERTY_NAME = "COLOR"
        }

        override fun getValue() = value?.name

        override fun setValue(name: String?) {
            name?.let {
                try {
                    value = EventColor.valueOf(name.toLowerCase())
                } catch(e: IllegalArgumentException) {
                    Constants.log.warning("Ignoring unknown COLOR $name")
                }
            }
        }

        override fun validate() {
        }

        class Factory: PropertyFactory<Color> {
            override fun createProperty() = Color()

            override fun createProperty(params: ParameterList?, value: String?): Color {
                val c = Color()
                c.setValue(value)
                return c
            }

            override fun supports(property: String?) = property == PROPERTY_NAME
        }

    }

    /** EMAIL parameter for ATTENDEE properties, as used by iCloud:
        ATTENDEE;EMAIL=bla@domain.tld;/path/to/principal
    */
    class Email(): Parameter(PARAMETER_NAME, ParameterFactoryImpl.getInstance()) {
        companion object {
            val FACTORY = Factory()
            val PARAMETER_NAME = "EMAIL"
        }

        var email: String? = null
        override fun getValue() = email

        constructor(aValue: String): this()
        {
            email = Strings.unquote(aValue)
        }

        class Factory: ParameterFactory<Email> {
            @Throws(URISyntaxException::class)
            override fun createParameter(value: String) = Email(value)

            override fun supports(name: String) = name == PARAMETER_NAME
        }
    }

}